### PR TITLE
ITE: drivers/serial: add the UART driver for the PM callback function

### PR DESCRIPTION
**Pull request by GTLin08**
IT8XXX2 uses shared ns16550.c driver which does not provide a power
management callback(pm_action_cb), so create driver to handle
IT8XXX2 specific UART features.

note: pm_action_cb(old name: pm_control_fn)

Change-Id: I4b32cd127f06151fabe99addbb2c7eceedc0b82d
Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

### CODE REVIEW

**Comment by dcpleung** [comment]
Is there any particular reason that you want to handle two UART devices in one driver instance? In Zephyr, we usually have one driver instance per device.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int uart_it8xxx2_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	int ret;
+
+#ifdef CONFIG_PM_DEVICE
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay)
```
Comment (keith-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:114]: This block will get called twice, once for UART1 and again for UART2.  You should add properties to the ite,it8xxx2-uart node to specify the gpio-dev and port-num, similar to what you did for the I2C driver.

Then you should be able to remove all the `#if DT_NODE_HAS_STATUS(DT_NODELABEL(uartN), okay)` code blocks.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

**Comment by GTLin08** [comment]
> Is there any particular reason that you want to handle two UART devices in one driver instance? In Zephyr, we usually have one driver instance per device.

UART1 is to access the EC(it8xxx2_evb) interface, and UART2 is used as the host interface and can be accessed by the host processor.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 1 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

**Comment by dcpleung** [comment]
> > Is there any particular reason that you want to handle two UART devices in one driver instance? In Zephyr, we usually have one driver instance per device.
> 
> UART1 is to access the EC(it8xxx2_evb) interface, and UART2 is used as the host interface and can be accessed by the host processor.

But they can be separated into two device instances, right?

**Comment by GTLin08** [comment]
> right

Yes, the previous code flow is indeed in the same instance. It's fixed now.

File: boards/riscv/it8xxx2_evb/it8xxx2_evb.dts [hunk 2]
```diff
@@ -78,6 +78,12 @@
 	current-speed = <460800>;
 	clock-frequency = <1804800>;
 };
+&ite_uart1_wrapper {
+	status = "okay";
+};
+&ite_uart1_wrapper {
+	status = "okay";
+};
```
Comment (brockus-zephyr) [inline boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:86]: why is this here twice?

Comment (GTLin08) [reply boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:86]: Fixed.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	/* UART handle */
+	const struct device *uart_dev;
+};
+
+enum uart_port_num {
+	UART1 = 1,
+	UART2,
+};
+
+#ifdef CONFIG_PM_DEVICE
```
Comment (keith-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:35]: If you add `depends PM_DEVICE` to the serial/Kconfig.it8xxx2 file, then I think you can remove the #ifdefs in this file.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	case PM_DEVICE_ACTION_SUSPEND:
+		if (config->port == UART1) {
+			/* Enable UART1 WUI */
+			ret = gpio_pin_interrupt_configure(config->gpio_dev, 0,
+							   GPIO_INT_TRIG_LOW);
+		} else if (config->port == UART2) {
+			/* Enable UART2 WUI */
+			ret = gpio_pin_interrupt_configure(config->gpio_dev, 1,
+							   GPIO_INT_TRIG_LOW);
+		}
```
Comment (keith-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:84]: I think you should add the GPIO pin number as a device tree property.

You can the eliminate this if branch.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_ite_it8xxx2.c (hunk 3 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c:33]: If add` depends PM_DEVICE` and CONFIG_PM_DEVICE is not enabled, uart_ite_it8xxx2.c will not be built.
But serial driver of ITE is still necessary, including the initialization of uart(will be moved here from soc.c later) and alternate of pinmux control(currently with hard code in pinmux.c).

File: drivers/serial/uart_ite_it8xxx2.c [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+#include <drivers/gpio.h>
+#include <drivers/uart.h>
+#include <kernel.h>
+#include <pm/device.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(uart_ite_it8xxx2, LOG_LEVEL_ERR);
+
+#define DRV_CONFIG(dev) \
```
[the remark this answers is not in the document]
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 4 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

**Comment by GTLin08** [comment]
Hi @keith-zephyr @brockus-zephyr @dcpleung,

I'd appreciate any comments of you.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+{
+	__unused const struct uart_it8xxx2_config *const config = dev->config;
+
+#ifdef CONFIG_PM_DEVICE
+	int ret = 0;
+
+	if (config->port == UART1) {
+		static struct gpio_callback uart1_wui_cb;
+
+		gpio_init_callback(&uart1_wui_cb, uart1_wui_isr, BIT(0));
```
Comment (keith-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:110]: Use `BIT(config->wakeup_pin)` in place of `BIT(0)`

File: drivers/serial/uart_ite_it8xxx2.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+		ret = gpio_add_callback(config->gpio_dev, &uart1_wui_cb);
+	} else if (config->port == UART2) {
+		static struct gpio_callback uart2_wui_cb;
+
+		gpio_init_callback(&uart2_wui_cb, uart2_wui_isr, BIT(1));
```
Comment (keith-zephyr) [inline drivers/serial/uart_ite_it8xxx2.c:116]: Same here - use `config->wakeup_pin`

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series [hunk 8]
```diff
@@ -26,6 +26,10 @@ config UART_NS16550_WA_ISR_REENABLE_INTERRUPT
 	default y
 	depends on UART_NS16550
 
+config UART_ITE_IT8XXX2
+	default y
+	depends on UART_NS16550
```
Comment (GTLin08) [inline soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:31]: This change also moves the UART config declare from the board level to the soc level.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 6 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 7 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: dts/riscv/it8xxx2.dtsi [hunk 9]
```diff
@@ -276,6 +276,28 @@
...
 		};
 
+		ite_uart1_wrapper: uartwrapper@f02720 {
+			compatible = "ite,it8xxx2-uart";
+			reg = <0x00f02720 0x0020>;
+			status = "disabled";
+			label = "UART1_WRAPPER";
+			port-num = <1>;
+			wakeup-pin = <0>;   /* GPB0 */
+			gpio-dev = <&gpiob>;
```
Comment (mbolivar-nordic) [inline dts/riscv/it8xxx2.dtsi:286]: Why isn't this using something like `wakeup-gpios = <&gpiob 0 ...>;`?

Comment (GTLin08) [reply dts/riscv/it8xxx2.dtsi:286]: You are right, we have planned to create another PR to add a WUI node in dtsi, and write the gpio node and pin to be configured like this. It will be more flexible to configure WUI for other drivers.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+#ifdef CONFIG_PM_DEVICE
+__weak void uart1_wui_isr_late(void) { }
+__weak void uart2_wui_isr_late(void) { }
+
+void uart1_wui_isr(const struct device *gpio, struct gpio_callback *cb,
+		   uint32_t pins)
+{
+	/* Disable interrupts on UART1 RX pin to avoid repeated interrupts. */
+	gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:41]: if return code is to be ignored make it explicit: `(void)func()`.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+	if (uart1_wui_isr_late) {
+		uart1_wui_isr_late();
+	}
+}
+
+void uart2_wui_isr(const struct device *gpio, struct gpio_callback *cb,
+		   uint32_t pins)
+{
+	/* Disable interrupts on UART2 RX pin to avoid repeated interrupts. */
+	gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:53]: ditto

File: drivers/serial/uart_ite_it8xxx2.c [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+				     GPIO_INT_DISABLE);
+
+	if (uart2_wui_isr_late) {
+		uart2_wui_isr_late();
+	}
+}
+
+static inline int uart_it8xxx2_pm_action_cb(const struct device *dev,
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:61]: by convention: `pm_action(`

File: drivers/serial/uart_ite_it8xxx2.c [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+					    enum pm_device_action action)
+{
+	const struct uart_it8xxx2_config *const config = dev->config;
+	int ret = 0;
+
+	switch (action) {
+	/* Next device power state is in active. */
+	case PM_DEVICE_ACTION_RESUME:
+		/* To do when EC is resuming. */
+		break;
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:71]: where is resume code?

File: drivers/serial/uart_ite_it8xxx2.c [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+				config->port, ret);
+			return ret;
+		}
+		/*
+		 * If uart device is busy with transmitting, the driver will
+		 * stay in while loop and wait for the transaction is completed.
+		 */
+		while (!uart_irq_tx_complete(config->uart_dev)) {
+			continue;
+		}
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:89]: PM callback should not be used to busy-wait until device finishes something that prevents the system to sleep (I assume it's the case). This means that the system will potentially start a power down transition, when in reality it can't. Consider using constraints API to prevent that (STM32 driver does that).

File: drivers/serial/uart_ite_it8xxx2.c [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int uart_it8xxx2_init(const struct device *dev)
+{
+	__unused const struct uart_it8xxx2_config *const config = dev->config;
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:102]: condition to `PM_DEVICE`, don't use __unused

File: drivers/serial/uart_ite_it8xxx2.c [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+				   BIT(config->wakeup_pin));
+
+		ret = gpio_add_callback(config->gpio_dev, &uart2_wui_cb);
+	}
+
+	if (ret) {
+		LOG_ERR("Failed to add UART%d callback (err %d)",
+			config->port, ret);
+		return ret;
+	}
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:127]: this looks like a gpio based wakeup, not really a wakeup based on e.g. uart activity, so I wonder if this can be made more generic

File: drivers/serial/uart_ite_it8xxx2.c [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+#endif /* CONFIG_PM_DEVICE */
+
+	return 0;
+}
+
+#define UART_ITE_IT8XXX2_INIT(inst)                                            \
+	static const struct uart_it8xxx2_config uart_it8xxx2_cfg_##inst = {    \
+		.port = DT_INST_PROP(inst, port_num),                          \
+		.wakeup_pin = DT_INST_PROP(inst, wakeup_pin),                  \
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:136]: you seem to be using the GPIO API, but DT just takes "pin", not something like `<&gpio0 1 <FLAGS>;`

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_ite_it8xxx2.c (hunk 16 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c:125]: Add a comment to explain that wakeup source is through UART Rx interrupt.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 14 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Removed, and constraints will be added to power_policy.c to determine. Thank for your information.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 15 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 17 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Fixed.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 13 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Fixed the comment.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 12 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 10 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 11 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: dts/riscv/it8xxx2.dtsi (hunk 9 above)
Comment (GTLin08) [reply dts/riscv/it8xxx2.dtsi]: This change has been fixed to `<&gpiob 0 ...>;`

File: drivers/serial/uart_ite_it8xxx2.c [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+	/* Next device power state is in active. */
+	case PM_DEVICE_ACTION_RESUME:
+		/* Nothing to do. */
+		break;
+	/* Next device power state is deep doze mode */
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Enable UART WUI */
+		ret = gpio_pin_interrupt_configure(config->gpio_dev,
+						   config->wakeup_pin,
+						   GPIO_INT_TRIG_LOW);
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:77]: does this action put UART into any sort of low-power mode?

File: drivers/serial/uart_ite_it8xxx2.c [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+	} else if (config->port == UART2) {
+		static struct gpio_callback uart2_wui_cb;
+
+		gpio_init_callback(&uart2_wui_cb, uart2_wui_isr,
+				   BIT(config->wakeup_pin));
+
+		ret = gpio_add_callback(config->gpio_dev, &uart2_wui_cb);
+	}
+
+	if (ret) {
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:121]: `ret < 0` (essentially boolean)

File: drivers/serial/uart_ite_it8xxx2.c [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+#endif /* CONFIG_PM_DEVICE */
+
+	return 0;
+}
+
+#define UART_ITE_IT8XXX2_INIT(inst)                                            \
+	static const struct uart_it8xxx2_config uart_it8xxx2_cfg_##inst = {    \
+		.port = DT_INST_PROP(inst, port_num),                          \
+		.gpio_dev = DEVICE_DT_GET(DT_INST_PHANDLE(inst, gpios)),       \
+		.wakeup_pin = DT_PHA(DT_DRV_INST(inst), gpios, pin),           \
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:135]: why can't `gpio_dt_spec` be used here?

File: drivers/serial/uart_ite_it8xxx2.c (hunk 14 above)
Comment (gmarull) [reply drivers/serial/uart_ite_it8xxx2.c]: the device is usually _who_ knows this information (e.g. pending transfer)

**Review by brockus-zephyr — APPROVED** [review]

File: drivers/serial/uart_ite_it8xxx2.c (hunk 18 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: The low power mode of it8xxx2 only has the case of PM_STATE_STANDBY(power.c). So, we need to configure the wakeup source before entering this mode.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 19 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 20 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Done.

File: drivers/serial/uart_ite_it8xxx2.c (hunk 14 above)
Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Ack.

File: drivers/serial/uart_ite_it8xxx2.c [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+
+	switch (action) {
+	/* Next device power state is in active. */
+	case PM_DEVICE_ACTION_RESUME:
+		/* Nothing to do. */
+		break;
+	/* Next device power state is deep doze mode */
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Enable UART WUI */
+		ret = gpio_pin_interrupt_configure(config->gpio_wui.port,
```
Comment (gmarull) [inline drivers/serial/uart_ite_it8xxx2.c:72]: please, use `gpio_pin_interrupt_dt_configure`

Comment (gmarull) [reply drivers/serial/uart_ite_it8xxx2.c:72]: have you considered using https://docs.zephyrproject.org/latest/doxygen/html/group__subsys__pm__sys.html#ga066945118b8f287ee56aacf41b677a78 ?

Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: Thank you for your suggestion, I will make changes in the future PR. 
But I want to know the difference between doing configuration here and doing pm_notifier_register(). Is the registration in this callback function?

Comment (GTLin08) [reply drivers/serial/uart_ite_it8xxx2.c]: > please, use `gpio_pin_interrupt_dt_configure`

Done.

**Review by keith-zephyr — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by gmarull — APPROVED** [review]
